sdk/core/Azure.Core/src/Pipeline/HttpPipelinePolicy.cs

```diff
@@ -46,5 +49,130 @@ protected static void ProcessNext(HttpMessage message, ReadOnlyMemory<HttpPipeli
         {
             pipeline.Span[0].Process(message, pipeline.Slice(1));
         }
+
+        /// <inheritdoc/>
+        public sealed override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+        {
+            if (message is not HttpMessage httpMessage)
+            {
+                throw new InvalidOperationException($"Invalid type for message: '{message?.GetType()}'");
+            }
+
+            if (pipeline is not HttpPipelineAdapter processor)
+            {
+                throw new InvalidOperationException($"Invalid type for pipeline: '{pipeline?.GetType()}'");
+            }
+
+            // If this method is called, it means this method is being called
+            // from a System.ClientModel PipelinePolicy instance. Since the
+            // contract for the pipeline parameter for Azure.Core and
+            // ClientModel Process methods is different, we need to pop some
+            // policies off the stack before calling Process on the Azure.Core policy.
+            await ProcessAsync(httpMessage, processor.Policies.Slice(currentIndex + 1)).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc/>
+        public sealed override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+        {
+            if (message is not HttpMessage httpMessage)
+            {
+                throw new InvalidOperationException($"Invalid type for message: '{message?.GetType()}'");
+            }
+
+            if (pipeline is not HttpPipelineAdapter processor)
+            {
+                throw new InvalidOperationException($"Invalid type for pipeline: '{pipeline?.GetType()}'");
+            }
+
+            // If this method is called, it means this method is being called
+            // from a System.ClientModel PipelinePolicy instance. Since the
+            // contract for the pipeline parameter for Azure.Core and
+            // ClientModel Process methods is different, we need to pop some
+            // policies off the stack before calling Process on the Azure.Core policy.
+            Process(httpMessage, processor.Policies.Slice(currentIndex + 1));
+        }
+
+        /// <summary>
+        /// This adapter adapts the Azure.Core <see cref="HttpPipeline"/> type's
```

Comment:
```suggestion
        /// This type adapts the Azure.Core <see cref="HttpPipeline"/>
```

nit: Phrasing was a bit awkward due to the repeated "adapter/adapts"